### ISSUE TITLE
Metadata On First Played Song When DJing Shows Unknown

### ISSUE THREAD

**Issue by CodeSteele** [AzuraCast/AzuraCast#867]
Apparently this is caused by LiquidSoap not firing the metadata change event that we expect to be fired.

Going to see what I can do about this...

**Comment by macbroadcast** [comment]
hey there @CodeSteele , I am using "butt" for live streaming, is this the reason the metadata info shown below isn't displayed either? 
![bildschirmfoto von 2018-10-18 22-09-22](https://user-images.githubusercontent.com/7961/47181173-9eb7e200-d322-11e8-8566-bb11118cf37d.png)

**Comment by CodeSteele** [comment]
@macbroadcast as far as I know, the metadata should be good after a single song plays. Does it not correct after the first song completes?

If not, likely a different issue.

**Comment by macbroadcast** [comment]
it does not change @CodeSteele so its useless for live DJs, what should i do ? Open an issue ?, buster seems always busy with less important stuff :smile: 
![bildschirmfoto von 2018-10-19 12-05-53](https://user-images.githubusercontent.com/7961/47212049-73230f00-d397-11e8-8822-f3ce2a40375b.png)

**Comment by CodeSteele** [comment]
@macbroadcast 

Oh that's right, BUTT just streams a single continuous stream...

Song metadata appears to be set here though:
![image](https://user-images.githubusercontent.com/31152835/47220000-752ca400-d37e-11e8-8cfa-17f9f2aee3df.png)

The delay I'm experiencing seems to match up with the delay in the audio stream though. So looks like it's fine.

Though I've had some issues where BUTT _sometimes_ doesn't update the song metadata correctly, so I'll loop back around and see if that is a BUTT problem or if it's just related to this bug once this is squashed.

**Comment by macbroadcast** [comment]
Thanks for your reply @CodeSteele , this works just with butt on ubuntu for me, will check settings on osx again and checking Mixx aswell!,  but where is the other information in BUTT for ? Description, URL Genre, IRC 
![bildschirmfoto von 2018-10-19 16-08-31](https://user-images.githubusercontent.com/7961/47223433-9068d500-d3b9-11e8-85a4-4d17eafa5b6e.png) Most of my DJs use serato and just plugin to BUTT with their stuff, so more information and metainfo is allways great since they play vinyl anyway :smile:

**Comment by CodeSteele** [comment]
Yeah not sure if BUTT supports that, may have to reach out to the BUTT devs -- I can't find anything in their docs.

**Comment by macbroadcast** [comment]
its seems its not capturing the information from butt on OSX https://stream.iradio.guru/public/iradio

**Comment by br0kenlink** [comment]
I'm having same problem with BUTT on live DJ streams, no metadata updating at all.
Also nicecast don't send anything.
On a side note, how do i connect to Icecast with Traktor  2? I can't seem to find user field in broadcasting tab in Traktor 2

**Comment by macbroadcast** [comment]
I am using butt for a long time, but never cared much about the metadata, I informed the author about the issue, because all my music stuff is of course on mac and therefore i  need a working solution, as well, but its mostly just a question of time. But I don't know what your problem with traktor is?https://traktorbible.com/en/t2_broadcasting_part_1.aspx#step2

**Comment by CodeSteele** [comment]
@macbroadcast as far as AzuraCast is concerned, it won't care if you're streaming from OSX or Ubuntu. If BUTT acts different between the two platforms I'd be _really_ tempted to blame BUTT for that.

Sadly right now neither of us have the ability to reproduce issues on OSX (complain to Apple about their EULA and vendor lock-in that doesn't allow us to spin up OSX VMs for testing, we do _a lot_ of our testing in VMs instead of having a ton of physical machines sitting around, which gets expensive fast).

Considering how slow BUTT development seems I should probably spend some time seeing if I can find a replacement, it has been a pain in our butt (heh) for awhile at this point due to a lot of odd behaviors/bugs.

**Comment by br0kenlink** [comment]
OK, no idea was it the cause but I now run latest AzuraCast in a docker and BUTT metadata works just fine sending from windows 10. Should know later if it works on OSX

**Comment by Vaalyn** [comment]
I've tested this on macOS High Sierra 10.13.6 (17G65) with Butt and Mixx and could only reproduce this bug with Butt. Also for me this Bug consistently occurs when starting Butt for the first time and then connecting. Butt sets the metadata to `ID3` when I just start and connect to AzuraCast. This also happened sometimes when manually setting the metadata in Butt.

```
I've looked at the station logs when this happens and there I found this:
2018/11/06 20:15:49 [lang:3] Authenticating DJ: shoutcast
2018/11/06 20:15:49 [test_live_fallback:3] Switch to audio_to_stereo_6052 with transition.
2018/11/06 20:15:49 [lang:3] AzuraCast DJ Auth Response: true
2018/11/06 20:15:49 [test_input_streamer:3] New metadata chunk ? -- ID3.
xmlEscapeEntities : char out of range
xmlEscapeEntities : char out of range
```

The problem most likely stems from an illegal char in the string that Butt sends to AzuraCast and that the `NowPlaying/Adapter/IceCast` can't properly decode in [this line](https://github.com/AzuraCast/nowplaying/blob/master/src/Adapter/Icecast.php#L98).

**Comment by Scraticus** [comment]
You can use Liquidsoap to capture the meta and change it if you like.

The way I had it on aerial was to do the following - feel free to adapt it for Azura:

`def fix_song(m) =
# The title metadata
  a = m["artist"]
  t = m["title"]
  l = string.split(separator=" -",t)
# If title is empty
  a = if ""==a then "PonyvilleFM" else a end
  a = if ""==a and list.length(l) >= 2 then t else a end
  t = if ""==t then "Missing Title!" else t end
# Otherwise if title = Unknown
  t = if "Unknown"==t then "A Wild DJ has Appeared!" else t end

  [("artist","#{a}"),("title","#{t}")]
end`
Some ogg encoders send a dead hunk at the beginning of a connection and this results in bad metadata being sent through LS and Icecast.

Back in 2012 when I wrote this, it was to stop IceCast from segfaulting and crashing the station. It's a simple fix that just sits in the script out of the way and fixes any bad meta.

Call it before after the DJ live functions on the LS script.

with :
`radio = map_metadata(fix_song,radio)`

**Comment by SlvrEagle23** [comment]
It appears that in the absence of any metadata, this just sets a placeholder value with the station name or other text. While we could implement this into the LS config, I'd prefer something a little more interactive, like triggering a metadata update upon successful DJ authentication that actually includes the DJ's name in the metadata. The [LS documentation](https://www.liquidsoap.info/doc-1.3.3/metadata.html) suggests this is possible via a Telnet connection, which I may take advantage of.

**Comment by SlvrEagle23** [comment]
While we can't do anything on the AzuraCast end about clients that connect to stream not sending an immediate metadata update, we can trigger one on our end when a DJ successfully authenticates, which overrides the existing metadata to show the DJ's "display name" and the station name.

That's exactly what the latest change accomplishes; it will flush the metadata upon a successful DJ connection, ensuring that your listeners are aware of the presence of a live DJ in the interstitial time between their connection and the first metadata push.

Hopefully this helps mitigate the issue!

**Comment by SlvrEagle23** [comment]
Reopening after having reverted the previous quick fix, as it causes the DJ name to show up in the song playback timeline and other places.

**Comment by gammaw** [comment]
Hi everyone, 

do you have an update on this? We have a perhaps [related issue](https://github.com/AzuraCast/AzuraCast/issues/1968) where Azuracast's JSON API shows inconsistent data after a live DJ connects via Butt. What I fail to see if it's a) an issue with Butt not sending certain data or sending wrong data to Azuracast or b) Azuracast not handling connection events properly (although I see that a) and b) are related). Our problem is that Azurcast seems to "remember" metadata accross different Butt sessions (carried out by different Butt clients).

Thanks

**Comment by aripisprecer** [comment]
Hello,

i am having this problem with the latest version of Azuracast on Docker (Debian 10 KVM).

I am streaming from Sam Broadcaster to Azuracast (SHOUTcast DNAS/posix(linux x64) v2.5.5.733 )

The streaming happends on a "LAN" network so there is no firewall, etc.

The stream is playing but on Azura web interface it shows it offline and from time to time it goes again online for 2-3 seconds.

Is there a way to fix this issue?

Thank you,
Marius

**Comment by aripisprecer** [comment]
Hello again,

i have figured it out: the problem was that i had 2 mount points.
The first one was a normal mount point and the second was configured a a relay to the first one (i don't know why).

I have deleted the relay mount point and now it display the station online and the song tile and artist are being displayed.

Sorry for the comment above.


Thank you!
Br
Marius

**Comment by gammaw** [comment]
If anyone interested a new verison of Butt is available (v0.1.19). I could only test it on Windows, the Mac version would not start for me due to some error. I see no difference compared to v0.1.18 in terms of the issue at hand. For me the behavior is deterministic: 
- After fresh Butt start: no song metadata is sent upon connection no matter how often I disconnect and reconnect.     
- After sending metadata at least once: the lastly entered metadata (confirmed with OK) is send upon each connect (until restarting Butt). 

(Although it's deterministic behavior, it's not really what we want. For us it would be great if it was configurable whether or not Butt automatically sends metadata or only "on demand".)

**Comment by JohnyBeGood** [comment]
> Hello again,
> 
> i have figured it out: the problem was that i had 2 mount points. The first one was a normal mount point and the second was configured a a relay to the first one (i don't know why).
> 
> I have deleted the relay mount point and now it display the station online and the song tile and artist are being displayed.
> 
> Sorry for the comment above.
> 
> Thank you! Br Marius

I know this is old post, but this is the only clue that got me to fix same "Stream Offline". The clue is match streaming software with relay name!
